### PR TITLE
Adding denies $KV.>/$OBJ.> along leaf connections on differing domain

### PR DESCRIPTION
**Pull request by matthiashanel**
Signed-off-by: Matthias Hanel <mh@synadia.com>

please read the introduced comment carefully

### CODE REVIEW

File: server/jetstream_cluster_test.go [hunk 1]
```diff
@@ -6875,7 +6875,8 @@ func TestJetStreamClusterDomainsWithNoJSHub(t *testing.T) {
 	// Client based API - Connected to the core cluster with no JS but account has JS.
 	s := c.randomServer()
 	// Make sure the JS interest from the LNs has made it to this server.
-	checkSubInterest(t, s, "NOJS", "$JS.SPOKE.API.>", time.Second)
+	checkSubInterest(t, s, "NOJS", "$JS.SPOKE.API.INFO", time.Second)
+	//checkSubInterest(t, s, "NOJS", "$JS.SPOKE.API.INFO", time.Second)
```
Comment (kozlovic) [inline server/jetstream_cluster_test.go:6879]: Remove that line?

File: server/leafnode.go [hunk 2]
```diff
@@ -1696,7 +1702,7 @@ func (c *client) forceAddToSmap(subj string) {
 	defer c.mu.Unlock()
 
 	n := c.leaf.smap[subj]
-	if n != 0 {
+	if n != 0 || c.leaf.smap == nil {
```
Comment (kozlovic) [inline server/leafnode.go:1705]: If would have personally do the check and return if that is the case, before the actual lookup.

File: server/leafnode_test.go [hunk 3]
```diff
@@ -5183,3 +5183,98 @@ func TestLeafNodeQueueGroupWithLateLNJoin(t *testing.T) {
...
+}
+system_account: SYS
+`
+
+	sd1 := createDir(t, JetStreamStoreDir)
+	defer os.RemoveAll(sd1)
+	confA := createConfFile(t, []byte(fmt.Sprintf(`
+listen: 127.0.0.1:-1
+%s
+jetstream: { domain: da, store_dir: %s, max_mem: 50Mb, max_file: 50Mb }
```
Comment (kozlovic) [inline server/leafnode_test.go:5201]: Please replace with `store_dir: '%s'`

File: server/leafnode_test.go [hunk 4]
```diff
@@ -5183,3 +5183,98 @@ func TestLeafNodeQueueGroupWithLateLNJoin(t *testing.T) {
...
+	defer removeFile(t, confA)
+	sA, _ := RunServerWithConfig(confA)
+	defer sA.Shutdown()
+
+	sd2 := createDir(t, JetStreamStoreDir)
+	defer os.RemoveAll(sd2)
+	confL := createConfFile(t, []byte(fmt.Sprintf(`
+listen: 127.0.0.1:-1
+%s
+jetstream: { domain: dl, store_dir: %s, max_mem: 50Mb, max_file: 50Mb }
```
Comment (kozlovic) [inline server/leafnode_test.go:5219]: here too

File: server/leafnode_test.go [hunk 5]
```diff
@@ -5183,3 +5183,98 @@ func TestLeafNodeQueueGroupWithLateLNJoin(t *testing.T) {
...
+	no_advertise: true
+    remotes:[{url:nats://a1:a1@127.0.0.1:%d, account: A},
+		     {url:nats://s1:s1@127.0.0.1:%d, account: SYS}]
+}
+`, accs, sd2, sA.opts.LeafNode.Port, sA.opts.LeafNode.Port)))
+	defer removeFile(t, confL)
+	sL, _ := RunServerWithConfig(confL)
+	defer sL.Shutdown()
+
+	ncA := natsConnect(t, sA.ClientURL(), nats.UserInfo("a1", "a1"))
```
Comment (kozlovic) [inline server/leafnode_test.go:5230]: Add `defer ncA.Close()`, and same for `ncL`

**Review by kozlovic — CHANGES_REQUESTED** [review]

File: server/jetstream_cluster_test.go (hunk 1 above)
Comment (matthiashanel) [reply server/jetstream_cluster_test.go:6879]: done

File: server/leafnode.go (hunk 2 above)
Comment (matthiashanel) [reply server/leafnode.go:1705]: not sure I can do that. the loop iterates over account clients

Comment (kozlovic) [reply server/leafnode.go:1705]: Just meant something like this:

Comment (matthiashanel) [reply server/leafnode.go:1705]: sorry misread that as loop not lookup. will fix

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**